elastic/kibana
# [data.search.searchSource] Update SearchSource to use Fields API.

Closes https://github.com/elastic/kibana/issues/77241
Fixes https://github.com/elastic/kibana/issues/22897
Fixes https://github.com/elastic/kibana/issues/75813
Unblocks https://github.com/elastic/kibana/issues/80517 & https://github.com/elastic/kibana/pull/83891
Relates to https://github.com/elastic/kibana/issues/69049
Prior experiments: https://github.com/elastic/kibana/pull/77348 https://github.com/elastic/kibana/pull/75407

## Summary

This PR integrates the new [search fields API](https://www.elastic.co/guide/en/elasticsearch/reference/master/search-fields.html#search-fields) with the high-level search service `SearchSource`.

One of the main problems encountered in https://github.com/elastic/kibana/pull/75407 was that there wasn't an ergonomic way to add support for passing an array of fields to the fields API as `SearchSource.setField('fields', ...);` already existed.

This PR updates the API for `.setField('fields', [...])` to pass values to the search fields API instead of retrieving values from source. For apps which want to keep the legacy behavior, `.setField('fieldsFromSource', [...])` has been added.

To make it easier to play around with this feature, the "Search Examples" sample plugin has been rewritten to incorporate SearchSource, and provides a way to easily configure your request in the UI and then see the query DSL & response that is returned based on your configuration.

![Screen Shot 2020-11-02 at 3 35 02 PM](https://user-images.githubusercontent.com/1608770/97926819-06ca9c80-1d21-11eb-8520-34b4c5b8fead.png)

## Behavior
Here's a rough summary of the new behavior:
- `setField('fields', [...])`
  - Source filters are applied first to remove any unneeded fields. Filters also get applied to stored/scripted fields.
  - Calls the search fields API
  - If you request any fields which are in the docvalues for `indexPattern.getComputedFields`, they will automatically be merged with the formats from the index pattern, unless you explicitly provide one.
  - You can pass `['*']` to get all fields
- `setField('fieldsFromSource', [...])`
  - Mimics legacy behavior in that it will request the specified fields from source via `_source.includes: [...]`
  - Note that docvalue fields have still been removed: those will now be requested from the fields API even when using this legacy behavior, and if you include a field that's in the index pattern's computed docvalues, it will be added to the fields API. But any other fields provided in the list will be pulled from source.

## Changes
- Added `fieldsFromSource` to SearchSource & refactored existing `fields` logic to use fields API
- I've updated apps that were previously using `fields` to use `fieldsFromSource` (Discover & Maps).
- Updated "source filters" UI in index patterns to call them "field filters" as they are no longer exclusively applied to `_source`
- Remove the [usage of docvalue_fields for formatting date fields](https://github.com/elastic/kibana/issues/75813) in favor of using the fields API, which [doesn't have the same limits as docvalues](https://github.com/elastic/kibana/issues/22897). While the changes to the request format are handled internally, the response that comes back from ES will be different, so this is still a breaking change for apps that use search source and rely on the docvalue_fields from the response. At this point Discover is the only app that I'm aware of which does this. The maps app is also using docvalues, but AFAICT they are manually adding them to the request & not relying on searchsource to "magically" handle it for them.

## Usage & Testing

1. Check out this PR
1. `yarn start --run-examples` & load some sample data
1. Navigate to the example plugin: `Developer examples > Search examples`
1. Specify desired query, filters, timerange, index pattern, & fields in the UI
1. Click "Request from high-level client"

## Open Questions
1. Do we want to apply source filters to the fields API at all? (The name implies the filters only get applied to `_source` itself so I'm worried this is confusing?)
    - I applied them to the fields API too and renamed source filters to "Field filters" in the UI. We should discuss if this is something that we are all comfortable with. (Edit: Agreed this makes the most sense and is what most users would expect)
1. Any other missing functionality that Discover would need to make this work? (So far the answer is "no")
1. Should we make the fields API opt-in/out, or always use it under the hood?
    - Decided to always use the fields API, and you must opt out with "fieldsFromSource"


cc @majagrubic @kertal @timroes @ppisljar @stacey-gammon 


## Plugin API Changes

### SearchSource now uses the search fields param by default
The `data` plugin's high-level search API, `SearchSource`, has migrated to use the [Elasticsearch search fields param](https://www.elastic.co/guide/en/elasticsearch/reference/7.x/search-fields.html#search-fields-param) as the default when constructing a search request body with specific fields. In order to make it as easy as possible for plugins to migrate to the new behavior, we've preserved a way for plugins to use the legacy behavior of requesting fields from `_source`:

```ts
class MyPlugin {
  start(core, { data }) {
    const searchSource = data.search.searchSource.create();

    // Deprecated. Legacy behavior from 'fields' has been moved to 'fieldsFromSource'.
    // This is now the only way to search for fields directly from `_source`:
    searchSource.setField('fieldsFromSource', ['fieldA', 'fieldB']);

    // The old 'fields' syntax now uses the search fields API under the hood, and accepts
    // an array of fields that are passed straight through to the fields API.
    searchSource.setField('fields', ['fieldC', { field: 'fieldD', format: 'date_time' });

    ...etc
  }
}
```

If your plugin relies on calling `setField('fields', [...])`, you should update it to use `fieldsFromSource` until you are able to adapt your plugin to the new fields behavior.

### SearchSource has stopped using `docvalue_fields` by default
Previously `SearchSource` would automatically request `docvalue_fields` for any date fields in an index pattern in order to avoid [a situation where Kibana might receive a date field from Elasticsearch that it doesn't know how to format](https://github.com/elastic/kibana/issues/22897#issuecomment-604998405). With the introduction of the [Elasticsearch search fields param](https://www.elastic.co/guide/en/elasticsearch/reference/7.x/search-fields.html#search-fields-param) which supports requesting fields in a particular format, we no longer need to rely on `docvalue_fields` for this behavior. `SearchSource` will now automatically request any date fields via the fields API, unless you provide specific ones via `setField('fields', [...])`, in which case only the relevant ones will be requested. If you do not provide a `format` for the fields you are requesting, one will automatically be added for you.
